jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -6616,36 +6620,205 @@ private static char convEndian(boolean isBigEndian, char value) {
+	 * Returns false if clz is an array class
+	 *
+	 * @param clz the the class to check
+	 * @param offset offset in bytes of the field in clz
+	 * @return a boolean indicating whether a flattened field exists at the given offset
+	 */
+	private native boolean isFieldAtOffsetFlattened(Class<?> clz, long offset);
+
+	/**
+	 * Sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if the comparison is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
+	private final <V> boolean compareAndSetFlattenedValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
+		boolean result = false;
+		if (getValue(obj, offset, clz) == v1) {
+			putValue(obj, offset, clz, v2);
+			result = true;
+		}
+		return result;
+	}
+
+	/**
+	 * Sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if the comparison is successful
+	 * @return value in obj at offset before this operation. This will be v1 if the exchange was successful
+	 */
+	private final <V> Object compareAndExchangeFlattenedValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
+		Object result = getValue(obj, offset, clz);
+		if (result == v1) {
+			putValue(obj, offset, clz, v2);
+		}
+		return result;
+	}
+
+	/**
+	 * Atomically sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of getVolatile.
+	 * The set operation has the memory semantics of setVolatile.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if the comparison is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
 	public final <V> boolean compareAndSetValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
-		throw OpenJDKCompileStubThrowError();
+		boolean result = false;
+		if (isFlattenedArray(obj.getClass()) || isFieldAtOffsetFlattened(obj.getClass(), offset)) {
+			synchronized (inlineTypesLock) {
+				result = compareAndSetFlattenedValue(obj, offset, clz, v1, v2);
+			}
+		} else {
+			result = compareAndSetObject(obj, offset, v1, v2);
```

Comment:
Better to use `compareAndSetReference()` which is the newer API.

Reply:
changed in e6eae18